### PR TITLE
[Rule tuning]  Correct tags with associated threat mappings

### LINKED ISSUE #1002: [Rule Tuning] Rule tags don't always match associated Tactics
## Description
When reviewing rules, we noticed the tags don't always align accurately with the associated Tactics. 

For example, a rule named `initial_access_unusual_dns_service_file_writes` should have `Initial Access` as a tag, instead of `Execution`.

See the example data below.  
This should be corrected for consistency and accuracy in the UI. 

## Rules needing update
```
  "e7075e8d-a966-458e-a183-85cd331af255 - Default Cobalt Strike Team Server Certificate -> expected: ['Threat Detection'] in ['Command and Control', 'Post-Execution', 'Threat Detection, Prevention and Hunting', 'Elastic', 'Network']",
  "45d273fb-1dca-457d-9855-bcb302180c21 - Encrypting Files with WinRar or 7z -> expected: ['Collection'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Exfiltration']",
  "e8571d5f-bea1-46c2-9f56-998de2d3ed95 - Local Service Commands -> expected: ['Lateral Movement'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "0e79980b-4250-4a50-a509-69294c14e84b - MsBuild Making Network Connections -> expected: ['Defense Evasion'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "b86afe07-0d98-4738-b15d-8d7465f95ff5 - Network Connection via MsXsl -> expected: ['Defense Evasion'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "a4ec1382-4557-452b-89ba-e413b22ed4b8 - Network Connection via Mshta -> expected: ['Defense Evasion'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "5cd55388-a19c-47c7-8ec4-f41656c2fded - Outbound Scheduled Task Activity via PowerShell -> expected: ['Execution'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Discovery', 'Lateral Movement']",
  "12f07955-1674-44f7-86b5-c35da0a6f41a - Suspicious Cmd Execution via WMI -> expected: ['Execution'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Lateral Movement']",
  "9a5b4e31-6cde-4295-9ff7-6be1b8567e1b - Suspicious Explorer Child Process -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "baa5d22c-5e1c-4f33-bfc9-efa73bb53022 - Suspicious Image Load (taskschd.dll) from MS Office -> expected: ['Persistence'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "a624863f-a70d-417f-a7d2-7a404638d47f - Suspicious MS Office Child Process -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "32f4675e-6c49-4ace-80f9-97c9259dca2e - Suspicious MS Outlook Child Process -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "852c1f19-68e8-43a6-9dce-340771fe1be3 - Suspicious PowerShell Engine ImageLoad -> expected: ['Execution'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Defense Evasion']",
  "66da12b1-ac83-40eb-814c-07ed1d82b7b9 - Suspicious macOS MS Office Child Process -> expected: ['Initial Access'] in ['Elastic', 'Host', 'macOS', 'Threat Detection', 'Execution']",
  "8c37dc0e-e3ac-4c97-8aa0-cf6a9122de45 - Unusual Child Process of dns.exe -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "c7ce36c0-32ff-4f9a-bfc2-dcb242bf99f9 - Unusual File Modification by dns.exe -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "52aaab7b-b51c-441a-89ce-4387b3aea886 - Unusual Network Connection via RunDLL32 -> expected: ['Defense Evasion'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "610949a1-312f-4e04-bb55-3a79b8c95267 - Unusual Process Network Connection -> expected: ['Defense Evasion'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "6a8ab9cc-4023-4d17-b5df-1a3e16882ce7 - Unusual Service Host Child Process - Childless Service -> expected: ['Privilege Escalation'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Defense Evasion']",
  "b5ea4bfe-a1b2-421f-9d47-22a75a6f2921 - Volume Shadow Copy Deletion via VssAdmin -> expected: ['Impact'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Defense Evasion']",
  "ec328da1-d5df-482b-866c-4a435692b1f3 - WPAD Service Exploit -> expected: ['Privilege Escalation'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "f545ff26-3c94-4fd0-bd33-3c7f95a3a0fc - Windows Script Executing PowerShell -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']",
  "b64b183e-1a76-422d-9179-7b389513e74d - Windows Script Interpreter Executing Process via WMI -> expected: ['Initial Access'] in ['Elastic', 'Host', 'Windows', 'Threat Detection', 'Execution']"
```

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
resolves #1002 

## Summary
Re-aligns tags with threat mappings and adds tests.

* ensures primary (0) tactic is in tags
* ensures any tactic tags are in threat mapping

<details>
<summary>Failure message</summary><br>

```
AssertionError: Rules with misaligned tags and tactics:
e6c98d38-633d-4b3e-9387-42112cd5ac10 - Authorization Plugin Modification ->
    unexpected (or missing from threat mapping): ['Credential Access']
e6c1a552-7776-44ad-ae0f-8746cc07773c - Bash Shell Profile Modification ->
    unexpected (or missing from threat mapping): ['Execution']
2edc8076-291e-41e9-81e4-e3fcbc97ae5e - Creation of a Hidden Local User Account ->
    unexpected (or missing from threat mapping): ['Defense Evasion']
45d273fb-1dca-457d-9855-bcb302180c21 - Encrypting Files with WinRar or 7z ->
    expected: ['Collection']
    unexpected (or missing from threat mapping): ['Exfiltration']
e8571d5f-bea1-46c2-9f56-998de2d3ed95 - Local Service Commands ->
    expected: ['Lateral Movement']
    unexpected (or missing from threat mapping): ['Execution']
483c4daf-b0c6-49e0-adf3-0bfa93231d6b - Microsoft Exchange Server UM Spawning Suspicious Processes ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
6cd1779c-560f-4b68-a8f1-11009b27fe63 - Microsoft Exchange Server UM Writing Suspicious Files ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
f81ee52c-297e-46d9-9205-07e66931df26 - Microsoft Exchange Worker Spawning Suspicious Processes ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
717f82c2-7741-4f9b-85b8-d06aeb853f4f - Modification of Dynamic Linker Preload Shared Object ->
    unexpected (or missing from threat mapping): ['Persistence']
0e79980b-4250-4a50-a509-69294c14e84b - MsBuild Making Network Connections ->
    expected: ['Defense Evasion']
    unexpected (or missing from threat mapping): ['Execution']
b86afe07-0d98-4738-b15d-8d7465f95ff5 - Network Connection via MsXsl ->
    expected: ['Defense Evasion']
    unexpected (or missing from threat mapping): ['Execution']
a4ec1382-4557-452b-89ba-e413b22ed4b8 - Network Connection via Mshta ->
    expected: ['Defense Evasion']
    unexpected (or missing from threat mapping): ['Execution']
5cd55388-a19c-47c7-8ec4-f41656c2fded - Outbound Scheduled Task Activity via PowerShell ->
    expected: ['Execution']
    unexpected (or missing from threat mapping): ['Discovery', 'Lateral Movement']
aa9a274d-6b53-424d-ac5e-cb8ca4251650 - Remotely Started Services via RPC ->
    unexpected (or missing from threat mapping): ['Execution']
12f07955-1674-44f7-86b5-c35da0a6f41a - Suspicious Cmd Execution via WMI ->
    expected: ['Execution']
    unexpected (or missing from threat mapping): ['Lateral Movement']
3e3d15c6-1509-479a-b125-21718372157e - Suspicious Emond Child Process ->
    unexpected (or missing from threat mapping): ['Execution']
9a5b4e31-6cde-4295-9ff7-6be1b8567e1b - Suspicious Explorer Child Process ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
baa5d22c-5e1c-4f33-bfc9-efa73bb53022 - Suspicious Image Load (taskschd.dll) from MS Office ->
    expected: ['Persistence']
    unexpected (or missing from threat mapping): ['Execution']
a624863f-a70d-417f-a7d2-7a404638d47f - Suspicious MS Office Child Process ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
32f4675e-6c49-4ace-80f9-97c9259dca2e - Suspicious MS Outlook Child Process ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
852c1f19-68e8-43a6-9dce-340771fe1be3 - Suspicious PowerShell Engine ImageLoad ->
    expected: ['Execution']
    unexpected (or missing from threat mapping): ['Defense Evasion']
66da12b1-ac83-40eb-814c-07ed1d82b7b9 - Suspicious macOS MS Office Child Process ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
8c37dc0e-e3ac-4c97-8aa0-cf6a9122de45 - Unusual Child Process of dns.exe ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
c7ce36c0-32ff-4f9a-bfc2-dcb242bf99f9 - Unusual File Modification by dns.exe ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
52aaab7b-b51c-441a-89ce-4387b3aea886 - Unusual Network Connection via RunDLL32 ->
    expected: ['Defense Evasion']
    unexpected (or missing from threat mapping): ['Execution']
610949a1-312f-4e04-bb55-3a79b8c95267 - Unusual Process Network Connection ->
    expected: ['Defense Evasion']
    unexpected (or missing from threat mapping): ['Execution']
6a8ab9cc-4023-4d17-b5df-1a3e16882ce7 - Unusual Service Host Child Process - Childless Service ->
    expected: ['Privilege Escalation']
b5ea4bfe-a1b2-421f-9d47-22a75a6f2921 - Volume Shadow Copy Deletion via VssAdmin ->
    expected: ['Impact']
    unexpected (or missing from threat mapping): ['Defense Evasion']
ec328da1-d5df-482b-866c-4a435692b1f3 - WPAD Service Exploit ->
    expected: ['Privilege Escalation']
    unexpected (or missing from threat mapping): ['Execution']
f545ff26-3c94-4fd0-bd33-3c7f95a3a0fc - Windows Script Executing PowerShell ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
b64b183e-1a76-422d-9179-7b389513e74d - Windows Script Interpreter Executing Process via WMI ->
    expected: ['Initial Access']
    unexpected (or missing from threat mapping): ['Execution']
```

</details>

### CODE REVIEW

File: rules/windows/privilege_escalation_unusual_svchost_childproc_childless.toml [hunk 1]
```diff
@@ -18,7 +18,7 @@ name = "Unusual Service Host Child Process - Childless Service"
 risk_score = 47
 rule_id = "6a8ab9cc-4023-4d17-b5df-1a3e16882ce7"
 severity = "medium"
-tags = ["Elastic", "Host", "Windows", "Threat Detection", "Defense Evasion"]
+tags = ["Elastic", "Host", "Windows", "Threat Detection", "Defense Evasion", "Privilege Escalation"]
```
Comment (bm11100) [inline rules/windows/privilege_escalation_unusual_svchost_childproc_childless.toml:21]:
```suggestion
tags = ["Elastic", "Host", "Windows", "Threat Detection", "Privilege Escalation"]
```

**Review by bm11100 — APPROVED** [review]
aligns with rule names, lgtm

File: rules/windows/privilege_escalation_unusual_svchost_childproc_childless.toml (hunk 1 above)
Comment (brokensound77) [reply rules/windows/privilege_escalation_unusual_svchost_childproc_childless.toml:21]: So I left the test (and automated change) to allow for more than 1 tactic, so long as it is defined in the threat mapping. Even though the norm is to just tag the primary tactic, I did not want to be too restrictive - unless we think we should limit to only a single tactic.

File: tests/test_all_rules.py [hunk 2]
```diff
@@ -292,6 +292,48 @@ def test_required_tags(self):
...
+    def test_primary_tactic_as_tag(self):
+        from detection_rules.attack import tactics
+
+        invalid = []
+        tactics = set(tactics)
+
+        for rule in self.rules:
+            rule_tags = rule.contents['tags']
+
+            if 'Continuous Monitoring' in rule_tags or rule.type == 'machine_learning':
```
Comment (rw-access) [inline tests/test_all_rules.py:304]: why do these get  special treatment?

File: tests/test_all_rules.py [hunk 3]
```diff
@@ -292,6 +292,48 @@ def test_required_tags(self):
...
+
+                if 'Threat Detection' not in rule_tags:
+                    missing.append('Threat Detection')
+
+                # missing primary tactic
+                if primary_tactic not in rule.contents['tags']:
+                    missing.append(primary_tactic)
+
+                # listed tactic that is not in threat mapping
+                tag_tactics = set(rule_tags) & tactics
```
Comment (rw-access) [inline tests/test_all_rules.py:321]: I struggle so much with the binary operators `&` and `|` for sets and all of the methods.
maybe spelling it out is the most clear?
```suggestion
                tag_tactics = set(rule_tags).intersection(tactics)
```

File: tests/test_all_rules.py [hunk 4]
```diff
@@ -292,6 +292,48 @@ def test_required_tags(self):
...
+                    if missing:
+                        err_msg += f'\n    expected: {missing}'
+                    if missing_from_threat:
+                        err_msg += f'\n    unexpected (or missing from threat mapping): {missing_from_threat}'
+
+                    invalid.append(err_msg)
+
+        if invalid:
+            err_msg = '\n'.join(invalid)
+            self.fail(f'Rules with misaligned tags and tactics:\n{err_msg}')
```
Comment (rw-access) [inline tests/test_all_rules.py:335]: we gotta get tactics completely out of tags in the first place.
it's dumb that we have to do this, and sign of a bad design.

 i think there are some new things unfolding that give us the ability  to search on non-tag fields

**Review by rw-access — APPROVED** [review]
LGTM. I think we'll be better when we don't have to duplicate tactics in tags. So I look forward to that.

File: tests/test_all_rules.py (hunk 2 above)
Comment (brokensound77) [reply tests/test_all_rules.py:304]: Most of the continuous monitoring rules are not inherently threat-centric rules, though they can certainly be mapped to a threat (in some instances). So early on, it was decided to leave it out of the tags. If we feel that has changed, we can update.

Am I missing anything @bm11100 ? I forget the context to ML

**Comment by brokensound77** [comment]
> LGTM. I think we'll be better when we don't have to duplicate tactics in tags. So I look forward to that.

Thanks for the review, and I agree with you there

**Pull request merged**